### PR TITLE
[Intel MKL] unit test for bfloat16 on matmul.

### PR DESCRIPTION
**Pull request by noim210**
[Intel MKL] unit test for bfloat16 on matmul.
Unit test for bfloat16 for matmul is created. This is the first PR for bfloat16 unit test creation. After getting reviews, rest of the mkl backed op's unit tests will be created following similar methods. 

In this unit tests, to evaluate the benchmark value we convert the twice. First float32 random values are generated. Later this values are converted to bfloat16 (which is used in the graph) and these bfloat16 values are converted back to float32 again for using the numpy library. (numpy and default gradient function doesn't have bfloat16 operations). 
If we don't convert twice and use the original random float32 values to get the benchmark results, we need to have very loose tolerance.

### CODE REVIEW

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 1]
```diff
@@ -63,9 +64,60 @@ def _GetTransposedMatrices(x, x_name, kwargs):
   else:
     return x
 
-
 class MatMulTest(test_lib.TestCase):
-  pass  # Filled in below
+  def test_bfloat16(self):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:68]: The main function already has a matmul testing template, for which it adds test cases for specific sizes and types.
https://github.com/tensorflow/tensorflow/blob/b2ed6eb035c841df8c78eef183616d8cd374c9fb/tensorflow/python/kernel_tests/matmul_op_test.py#L259

We shouldn't replicate the logic. Can we add the bf16 data type to the `dtype_to_test` list instead?
https://github.com/tensorflow/tensorflow/blob/b2ed6eb035c841df8c78eef183616d8cd374c9fb/tensorflow/python/kernel_tests/matmul_op_test.py#L230

We can add `if`s for things specific to bf16 in `_GetMatMulTest` (the template).
https://github.com/tensorflow/tensorflow/blob/b2ed6eb035c841df8c78eef183616d8cd374c9fb/tensorflow/python/kernel_tests/matmul_op_test.py#L71

For example, this checks for fp16
https://github.com/tensorflow/tensorflow/blob/b2ed6eb035c841df8c78eef183616d8cd374c9fb/tensorflow/python/kernel_tests/matmul_op_test.py#L78-L81

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 2]
```diff
@@ -108,10 +160,59 @@ def Test(self):
 
   return Test
 
-
 class MatMulGradientTest(test_lib.TestCase):
-  pass  # Will be filled in below.
-
+  def test_Gradient_bfloat16(self):
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:164]: Same comment. We can modify the template function below for bf16-specific actions:
https://github.com/tensorflow/tensorflow/blob/b2ed6eb035c841df8c78eef183616d8cd374c9fb/tensorflow/python/kernel_tests/matmul_op_test.py#L116

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR!

File: tensorflow/python/kernel_tests/matmul_op_test.py (hunk 1 above)
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py:68]: Thanks , I will look into it

**Comment by gbaned** [comment]
@noim210  Can you please check @penpornk's comments and keep us posted ? Thanks!

File: tensorflow/python/kernel_tests/matmul_op_test.py (hunk 1 above)
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: I have updated the coded according to your suggestion.

File: tensorflow/python/kernel_tests/matmul_op_test.py (hunk 2 above)
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: I have updated the code according to your suggestion.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 3]
```diff
@@ -52,7 +53,12 @@ def _AddTest(test, op_name, testcase_name, fn):
   test_name = "_".join(["test", op_name, testcase_name])
   if hasattr(test, test_name):
     raise RuntimeError("Test %s defined more than once" % test_name)
-  setattr(test, test_name, test_util.deprecated_graph_mode_only(fn))
+  if 'bfloat16' in testcase_name:
+    # running bfloat16 in eager mode as we need to calculate intermediate
+    # tensors by using cast operators
+    setattr(test, test_name, fn)
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:59]: I think this won't be necessary once we use `.as_numpy_dtype`.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 4]
```diff
@@ -63,29 +69,42 @@ def _GetTransposedMatrices(x, x_name, kwargs):
...
 
 def _GetMatMulTest(a_np_, b_np_, use_static_shape_, **kwargs_):
 
   @test_util.run_without_tensor_float_32("Tests matmul")
   def Test(self):
-    np_val = np.matrix(a_np_) * np.matrix(b_np_)
+    if a_np_.dtype is dtypes.bfloat16:
+      # converting it back to float32 to avoid precision error
+      a_np_fp32 = np.array(math_ops.cast(a_np_, dtypes.float32))
+      b_np_fp32 = np.array(math_ops.cast(b_np_, dtypes.float32))
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:82]: Once we use `.as_numpy_dtype`, we can just call `as_type` here.
```suggestion
      a_np_fp32 = a_np_.as_type(np.float32)
      b_np_fp32 = b_np_.as_type(np.float32)
```
Same for Gradient.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 5]
```diff
@@ -63,29 +69,42 @@ def _GetTransposedMatrices(x, x_name, kwargs):
...
-  pass  # Filled in below
-
+  pass
 
 def _GetMatMulTest(a_np_, b_np_, use_static_shape_, **kwargs_):
 
   @test_util.run_without_tensor_float_32("Tests matmul")
   def Test(self):
-    np_val = np.matrix(a_np_) * np.matrix(b_np_)
+    if a_np_.dtype is dtypes.bfloat16:
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:79]:
```suggestion
    if a_np_.dtype is dtypes.bfloat16.as_numpy_dtype:
```

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 6]
```diff
@@ -108,37 +127,51 @@ def Test(self):
...
+    else:
+      # Transpose and possibly conjugate a_np_ and b_np_ according to the
+      # attributes such that tf.matmul(effective_a_np, effective_b_np, **kwargs)
+      # results in a valid matrix multiplication and produces the same result as
+      # np.matrix(a_np_) * np.matrix(b_np_)
+      effective_a_np = _GetTransposedMatrices(a_np_, "a", kwargs_)
+      effective_b_np = _GetTransposedMatrices(b_np_, "b", kwargs_)
+      effective_a_np_ = effective_a_np
+      effective_b_np_ = effective_b_np
+      epsilon = np.finfo(a_np_.dtype).eps
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:161]: This line can be taken off of `if/else`.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 7]
```diff
@@ -108,37 +127,51 @@ def Test(self):
...
+    if a_np_.dtype is dtypes.bfloat16:
+      # converting it back to float32 to avoid precision error
+      a_np_fp32 = np.array(math_ops.cast(a_np_, dtypes.float32))
+      b_np_fp32 = np.array(math_ops.cast(b_np_, dtypes.float32))
+      # Applying GetTransposedMatrix from random inputs for feeding to graph
+      a_np_trans = _GetTransposedMatrices(a_np_fp32, "a", kwargs_)
+      b_np_trans = _GetTransposedMatrices(b_np_fp32, "b", kwargs_)
+      # converting the Transpose/Adjoint matrix to bfloat16
+      effective_a_np = math_ops.cast(a_np_trans, dtypes.bfloat16)
+      effective_b_np = math_ops.cast(b_np_trans, dtypes.bfloat16)
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:148]:
```suggestion
      effective_a_np = a_np_trans.as_type(a_np_.dtype)
      effective_b_np = b_np_trans.as_type(b_np_.dtype)
```

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 8]
```diff
@@ -108,37 +127,51 @@ def Test(self):
...
+      effective_a_np_ = a_np_trans
+      effective_b_np_ = b_np_trans
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:150]: We can delete `effective_a_np_` and `effective_b_np_` from both `if/else`.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fixes! I think we can use the property `as_numpy_dtype` to help simplify the code.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 9]
```diff
@@ -265,6 +309,8 @@ def testInfixMatmulDoesDotProduct(self):
                              transpose_a=transpose_a,
                              adjoint_b=adjoint_b,
                              transpose_b=transpose_b))
+                #if dtype is dtypes.bfloat16:
+                #    continue;
```
[the remark this answers is not in the document]
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py:313]: yes it is. Thank you for pointing out. I will remove that

File: tensorflow/python/kernel_tests/matmul_op_test.py (hunk 3 above)
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: Unfortunately, this is again necessary as while using compute_gradient(fn,x,delta) function, we can't send x as bfloat16, it is giving us wrong numerical values (mainly zero values). So we need to send x as float32.  [more details in the response to comments below]. Using the decorator can't allow to use float32 for x.

File: tensorflow/python/kernel_tests/matmul_op_test.py (hunk 5 above)
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: Thanks, changed as you mentioned.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 10]
```diff
@@ -63,29 +69,42 @@ def _GetTransposedMatrices(x, x_name, kwargs):
...
   @test_util.run_without_tensor_float_32("Tests matmul")
   def Test(self):
-    np_val = np.matrix(a_np_) * np.matrix(b_np_)
+    if a_np_.dtype is dtypes.bfloat16:
+      # converting it back to float32 to avoid precision error
+      a_np_fp32 = np.array(math_ops.cast(a_np_, dtypes.float32))
+      b_np_fp32 = np.array(math_ops.cast(b_np_, dtypes.float32))
+      # calculating the benchmark values
+      np_val_temp = np.matrix(a_np_fp32) * np.matrix(b_np_fp32)
+      np_val = math_ops.cast(np_val_temp, dtypes.bfloat16)
```
[the remark this answers is not in the document]
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: Thanks, changed according to your suggestion.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 11]
```diff
@@ -63,29 +69,42 @@ def _GetTransposedMatrices(x, x_name, kwargs):
...
+      # Applying GetTransposedMatrix from random inputs for feeding to graph
+      a_np_trans = _GetTransposedMatrices(a_np_fp32, "a", kwargs_)
+      b_np_trans = _GetTransposedMatrices(b_np_fp32, "b", kwargs_)
+      # converting the Transpose/Adjoint matrix to bfloat16
+      effective_a_np = math_ops.cast(a_np_trans, dtypes.bfloat16)
+      effective_b_np = math_ops.cast(b_np_trans, dtypes.bfloat16)
```
[the remark this answers is not in the document]
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: changed like you said.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 12]
```diff
@@ -108,37 +127,51 @@ def Test(self):
 
   return Test
 
-
 class MatMulGradientTest(test_lib.TestCase):
-  pass  # Will be filled in below.
-
+  pass
```
[the remark this answers is not in the document]
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: done

File: tensorflow/python/kernel_tests/matmul_op_test.py (hunk 7 above)
Comment (noim210) [reply tensorflow/python/kernel_tests/matmul_op_test.py]: Thanks . changed.

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 13]
```diff
@@ -70,25 +70,25 @@ def _GetTransposedMatrices(x, x_name, kwargs):
     return x
 
 class MatMulTest(test_lib.TestCase):
-  pass
+  pass # Filled in below
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:73]: Nit: Please put two spaces between the code (`pass`) and the comment (`#`), i.e.,
```suggestion
  pass  # Filled in below
```
Per our [Python style guide](https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments).

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 14]
```diff
@@ -128,24 +128,24 @@ def Test(self):
   return Test
 
 class MatMulGradientTest(test_lib.TestCase):
-  pass
+  pass # Filled in below
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:131]: Nit: Please put two spaces between the code (`pass`) and the comment (`#`), i.e.,
```suggestion
  pass  # Filled in below
```
Per our [Python style guide](https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments).

File: tensorflow/python/kernel_tests/matmul_op_test.py [hunk 15]
```diff
@@ -70,25 +70,25 @@ def _GetTransposedMatrices(x, x_name, kwargs):
...
 class MatMulTest(test_lib.TestCase):
-  pass
+  pass # Filled in below
 
 def _GetMatMulTest(a_np_, b_np_, use_static_shape_, **kwargs_):
 
   @test_util.run_without_tensor_float_32("Tests matmul")
   def Test(self):
-    if a_np_.dtype is dtypes.bfloat16:
+    if a_np_.dtype==dtypes.bfloat16.as_numpy_dtype:
```
Comment (penpornk) [inline tensorflow/python/kernel_tests/matmul_op_test.py:79]: Nit: Please add a space on each side of `==`, i.e.,
```suggestion
    if a_np_.dtype == dtypes.bfloat16.as_numpy_dtype:
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes and the explanation! I have a few minor formatting nits and we should be done. :)

**Review by penpornk — APPROVED** [review]
Thank you for your patience!

**Review by gbaned — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by gbaned** [comment]
@noim210 Can you please resolve conflicts? Thanks!

**Pull request closed without merging**